state/core.go

```diff
@@ -199,6 +227,71 @@ func (c *Core) Height(commit flow.StateCommitment) (uint64, error) {
 	return height, nil
 }
 
+func (c *Core) Events(height uint64, types ...string) ([]flow.Event, error) {
+	// Make sure that the request is for a height below the currently active
+	// sentinel height; otherwise, we haven't indexed yet and we might return
+	// false information.
+	if height > c.height {
+		return nil, fmt.Errorf("unknown height (current: %d, requested: %d)", c.height, height)
+	}
+
+	lookup := make(map[uint64]struct{})
+	for _, typ := range types {
+		lookup[xxhash.Checksum64([]byte(typ))] = struct{}{}
+	}
+
+	// Iterate over all keys within the events index which are prefixed with the right block height.
+	var events []flow.Event
+	prefix := make([]byte, 1+8)
+	prefix[0] = model.PrefixEventIndex
+	binary.BigEndian.PutUint64(prefix[1:], height)
+	err := c.index.View(func(tx *badger.Txn) error {
+		it := tx.NewIterator(badger.IteratorOptions{
+			Prefix: prefix,
+		})
+		defer it.Close()
+
+		// Iterate on all keys with the right prefix.
+		for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+			item := it.Item()
+			key := item.Key()
+
+			// If types were given for filtering, discard events which should not be included.
+			hash := binary.BigEndian.Uint64(key[1+8:])
+			_, ok := lookup[hash]
+			if len(lookup) != 0 && !ok {
+				continue
```

Comment:
It might be worth to seek forward to the next key here, so that we don't read tons of events of the same type that we all ignore. You could do `hash++` (good thing it's already a `uint64`!) and then `it.Seek(prefix)` with the new hash put into the right place.

Reply:
What do you mean? When we `continue` here it is going to loop again, and the loop is

```go
for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() { }
```

so it is going to seek again, no? 🤔